ui/app/hooks/useTransactionDisplayData.js

```diff
@@ -0,0 +1,152 @@
+import { useTokenData } from './useTokenData'
+import { tokenSelector } from '../selectors'
+
+// This is duplicated from transactions selectors, additionally
+// from the history view we will already know the pending status
+// of a transaction, but in order to make this hook less coupled
+// to where it is used we should determine the pending status here
+const pendingStatusMap = {
+  [APPROVED_STATUS]: true,
+  [SUBMITTED_STATUS]: true,
+}
+
+/**
+ * @typedef {Object} TransactionDisplayData
+ * @property {string} title               - primary description of the transaction
+ * @property {string} subtitle            - supporting text describing the transaction
+ * @property {string} category            - the transaction category
+ * @property {string} primaryCurrency     - the currency string to display in the primary position
+ * @property {string} [secondaryCurrency] - the currency string to display in the secondary position
+ * @property {string} status              - the status of the transaction
+ * @property {string} senderAddress       - the Ethereum address of the sender
+ * @property {string} recipientAddress    - the Ethereum address of the recipient
+ */
+
+/**
+ * useTransactionDisplayData
+ *
+ * The goal of this method is to perform all of the necessary computation and
+ * state access required to take a transactionGroup and derive from it a shape
+ * of data that can power all views related to a transaction. Presently the main
+ * case is for shared logic between transaction-list-item and transaction-detail-view
+ * @param {Object} transactionGroup group of transactions
+ * @return {TransactionDisplayData}
+ */
+export function useTransactionDisplayData (transactionGroup) {
+  const knownTokens = useSelector(tokenSelector)
+  const t = useI18nContext()
+  const { initialTransaction, primaryTransaction } = transactionGroup
+  // initialTransaction contains the data we need to derive the primary purpose of this transaction group
+  const { transactionCategory } = initialTransaction
+
+  const { from: senderAddress, to } = initialTransaction.txParams || {}
+
+  // for smart contract interactions, methodData can be used to derive the name of the action being taken
+  const methodData = useSelector((state) => getKnownMethodData(state, initialTransaction?.txParams?.data)) || {}
+
+  const actionKey = getTransactionActionKey(initialTransaction)
+  const statusKey = getStatusKey(primaryTransaction)
+  const status = statusKey in pendingStatusMap ? PENDING_STATUS : statusKey
+
+  const primaryValue = primaryTransaction.txParams?.value
+  let prefix = '-'
+  const date = formatDateWithYearContext(initialTransaction.time || 0)
+  let subtitle
+  let recipientAddress = to
+
+  const token = knownTokens.find((token) => token.address === recipientAddress)
+  const tokenData = useTokenData(initialTransaction?.txParams?.data)
+  const tokenDisplayValue = useTokenDisplayValue(initialTransaction?.txParams?.data, token)
```

Comment:
The problem is you can't use hooks inside logic blocks, they have to be top level.

Reply:
That is why the useTokenData and useTokenDisplayValue hooks care for the cases in which they are not needed by returning null at the top.

Reply:
Ah, right. Using them could certainly be contingent upon the transaction being a token transfer though.

Reply:
`tokenData` is only used in the case of the transaction being a `transfer` or `transferfrom` category. `tokenDisplayValue` is passed to the `useCurrencyDisplay` hooks which also have to be top-level, but if `tokenDisplayValue` is undefined/null it won't be used by that hook.

Reply:
`tokenDisplayValue` might be truthy for transactions that aren't token transfers though. In those cases, the `primaryCurrency` would be displayed incorrectly.

Maybe I made this comment on the wrong line - this might be better solved by adding some conditions to whether these are passed to `useCurrencyDisplay`

Reply:
I think this is now resolved by explicitly passing undefined when not dealing with a token transfer transactionCategory